org/mozilla/jss/ssl/javax/JSSSocket.java

```diff
@@ -0,0 +1,869 @@
+ */
+public class JSSSocket extends SSLSocket {
+    /**
+     * Name of the SSLEngine protocol to use.
+     */
+    private String engineProviderProtocol = "TLS";
+
+    /**
+     * Name of the SSLEngine provider to use.
+     */
+    private String engineProvider = "Mozilla-JSS";
+
+    /**
+     * SSLContext to use to create the JSSEngine. Note that JSSSocket will fail
+     * if the context doesn't create JSSEngine instances.
+     */
+    private SSLContext jssContext;
+
+    /**
+     * JSSEngine instance to utilize for SSLEngine operations.
+     */
+    private JSSEngine engine;
+
+    /**
+     * All registered handshake callbacks.
+     */
+    private ArrayList<HandshakeCompletedListener> handshakeCallbacks = new ArrayList<HandshakeCompletedListener>();
+
+    /**
+     * The socket this JSSSocket was created over; all read/write operations
+     * go through this socket and all information exposed via Socket members
+     * go through here.
+     */
+    private Socket parent;
+
+    /**
+     * Previously consumed data, if any; utilized for certain SSLSocketFactory
+     * calls.
+     */
+    private InputStream consumedData;
+
+    /**
+     * Underlying SocketChannel for this socket; always exists.
+     */
+    private JSSSocketChannel channel;
+
+    /**
+     * Whether or not to automatically close the underlying Socket when this
+     * socket has been closed; defaults to true.
+     */
+    private boolean autoClose = true;
+
+    /**
+     * Start building a new JSSSocket.
+     *
+     * We specifically avoid creating any other constructors as we wish to
+     * consume an existing socket rather than creating a new one.
+     */
+    public JSSSocket() {}
+
+    /**
+     * Consume a parent socket, utilizing it for all read/write operations.
+     *
+     * This JSSSocket instance will inherit all information about the
+     * connection from this underlying socket. When utilized in a JSSSocket,
+     * callers should refrain from interacting with the underlying socket
+     * directly until the TLS connection is closed. Otherwise, messages may
+     * get dropped.
+     */
+    public void consumeSocket(Socket parent) throws IOException {
+        if (parent == null) {
+            String msg = "Unable to consume and utilize null parent socket!";
+            throw new IOException(msg);
+        }
+
+        this.parent = parent;
+    }
+
+    /**
+     * Get the SSLContext if one exists or create a new instance.
+     *
+     * This is used by initSSLEngine(..) to create the underlying SSLEngine.
+     */
+    protected SSLContext getSSLContext() throws IOException {
+        if (jssContext == null) {
+            try {
+                jssContext = SSLContext.getInstance(engineProviderProtocol, engineProvider);
+            } catch (Exception e) {
+                String msg = "Unable to create JSSSocket prior to Mozilla-JSS ";
+                msg += "initialization! " + e.getMessage();
```

Comment:
I ended up not really using this method. There's two issues with it:

 1. We have no real way of _initializing_ a JSSContext is one doesn't exist,
 2. We have no other way of creating `SSLEngine`'s. 

The problem is that, if we don't pass an explicit `SSLContext` instance (backed by our `JSSContextSpi`), we need to initialize it somehow. That means divining `KeyManager`s and `TrustManager`s to use.

Since we have a split `JSSEngine` interface/implementation (and `JSSContextSpi` currently controls which implementation gets selected -- where currently `JSSEngineReferenceImpl` is the only one available), we need _some_ context to use here.

When we construct a `JSSSocketFactory` from _within_ `JSSContextSpi` we additionally don't have a way of accessing the _parent_ `SSLContext` of this `JSSContextSpi` either! So we end up creating a new one! We could perhaps use the `SSLContext(SSLContextSpi, String, String)` constructor though... but we'd need to divine the `protocol` parameter still....

Thoughts for avoiding this problem definitely appreciated. :-)